### PR TITLE
feat: download insights raw data as csv

### PR DESCRIPTION
**Pull request by alannnc**
## What does this PR do?

- Adds button to insights that triggers new api endpoint that return file download of same data used for all the insights graphs.

Fixes #10747 

image

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How should this be tested?

- Are there environment variables that should be set? No
- What are the minimal test data to have? using yarn seed-insights from packages/playwright would be useful to have an entire insights setup with bookings/users/event-types ready.
- What is expected (happy path) to have (input and output)? Download data that was used for the same count for the graphs
- Any other important info that could help to test that PR

## Mandatory Tasks

- [X] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

File: packages/prisma/migrations/20231001101010_add_user_email_to_booking_time_status/migration.sql [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            WHEN "Booking"."endTime" < now() THEN 'completed'::text
+            WHEN "Booking"."endTime" > now() THEN 'uncompleted'::text
+            ELSE NULL::text
+        END AS "timeStatus",
+    et."parentId" AS "eventParentId",
+    "u"."email" AS "userEmail",
+    "u"."username" AS "username"
+   FROM "Booking"
+     LEFT JOIN "EventType" et ON "Booking"."eventTypeId" = et.id
+     LEFT JOIN users u ON u.id = "Booking"."userId";
```
Comment (alannnc) [inline packages/prisma/migrations/20231001101010_add_user_email_to_booking_time_status/migration.sql:34]: Removed unused membership left join here and added new users to get user->email and username

File: apps/web/pages/api/insights/download.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      startDate,
+      endDate,
+      teamId,
+      userId,
+      memberUserId,
+      isAll,
+    });
+
+    const csvAsString = objectToCsv(csvData);
+    const downloadAs = `Insights-${dayjs(startDate).format("YYYY-MM-DD")}-${dayjs(endDate).format(
```
Comment (alannnc) [inline apps/web/pages/api/insights/download.ts:78]: Felt that this name convention was right.

File: apps/web/pages/api/insights/download.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  }
+};
+
+function objectToCsv(data: Record<string, unknown>[]) {
+  // if empty data return empty string
+  if (!data.length) {
+    return "";
+  }
+  const header = Object.keys(data[0]).join(",") + "\n";
+  const rows = data.map((obj: any) => Object.values(obj).join(",") + "\n");
```
Comment (alannnc) [inline apps/web/pages/api/insights/download.ts:100]: returning props of object as first line for headers.

File: apps/web/playwright/insights.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
+import { expect } from "@playwright/test";
+
+import { randomString } from "@calcom/lib/random";
+import prisma from "@calcom/prisma";
+
+import { test } from "./lib/fixtures";
+
+test.describe.configure({ mode: "parallel" });
+
+const createTeamsAndMembership = async (userIdOne: number, userIdTwo: number) => {
```
Comment (alannnc) [inline apps/web/playwright/insights.e2e.ts:10]: There is another PR that have more test here, reused some logic but only added required test here.

File: packages/features/insights/server/events.ts [hunk 5]
```diff
@@ -213,6 +214,167 @@ class EventsInsights {
...
+        title: true,
+        createdAt: true,
+        timeStatus: true,
+        eventTypeId: true,
+        eventLength: true,
+        startTime: true,
+        endTime: true,
+        paid: true,
+        userEmail: true,
+        username: true,
```
Comment (alannnc) [inline packages/features/insights/server/events.ts:237]: We should be fine for now for this values.

File: packages/features/insights/server/events.ts [hunk 6]
```diff
@@ -213,6 +214,167 @@ class EventsInsights {
...
+   * - userId
+   * - teamId
+   * Generics will be:
+   * - isAll
+   * - startDate
+   * - endDate
+   * @param props
+   * @returns
+   */
+  static obtainWhereConditional = async (
```
Comment (alannnc) [inline packages/features/insights/server/events.ts:260]: Reusing logic from insights router, to be used later and avoid duplicated code.

File: packages/features/insights/server/events.ts [hunk 7]
```diff
@@ -213,6 +214,167 @@ class EventsInsights {
...
+  };
+
+  /*
+   * This is meant to be used for all functions inside insights router, ideally we should have a view that have all of this data
+   * The order where will be from the most specific to the least specific
+   * starting from the top will be:
+   * - memberUserId
+   * - eventTypeId
+   * - userId
+   * - teamId
```
Comment (alannnc) [inline packages/features/insights/server/events.ts:252]: Important to know.

**Comment by PeerRich** [comment]
getting a real-time URL comes later, right?

**Comment by PeerRich** [comment]
getting this:

CleanShot 2023-10-02 at 12 52 43@2x

on https://dev-2bibpkto5-cal-staging.vercel.app/insights

**Comment by alannnc** [comment]
Note: This won't work on preview urls since it requires migrations.

File: packages/features/insights/filters/Download/index.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    destinationUrl.searchParams.set("endDate", filter.dateRange[1].toISOString());
+    filter.selectedTeamId && destinationUrl.searchParams.set("teamId", filter.selectedTeamId?.toString());
+    filter.selectedUserId && destinationUrl.searchParams.set("userId", filter.selectedUserId?.toString());
+    filter.selectedEventTypeId &&
+      destinationUrl.searchParams.set("eventType", filter.selectedEventTypeId?.toString());
+    filter.selectedMemberUserId &&
+      destinationUrl.searchParams.set("memberUserId", filter.selectedMemberUserId?.toString());
+
+    // Simulate a loading delay (replace this with your actual download preparation logic)
+    setTimeout(() => {
```
Comment (CarinaWolli) [inline packages/features/insights/filters/Download/index.tsx:31]: Do we still need to replace that?

File: apps/web/pages/api/insights/download.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    const session = await getServerSession({ req, res });
+
+    if (!session?.user?.id) {
+      res.status(401).json({ message: "Not authenticated" });
+      return;
+    }
+
+    // Determine if the user is allowed to access the data and what role
+    if (isAll && !!!session.user.org) {
```
Comment (CarinaWolli) [inline apps/web/pages/api/insights/download.ts:62]:
```suggestion
    if (isAll && !session.user.org) {
```

File: apps/web/pages/api/insights/download.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    res.status(500).json({ message: "Internal server error" });
+    return;
+  }
+  try {
+    const { startDate, endDate, teamId, userId, memberUserId, isAll } = data;
+
+    const session = await getServerSession({ req, res });
+
+    if (!session?.user?.id) {
```
Comment (CarinaWolli) [inline apps/web/pages/api/insights/download.ts:56]: Should we also check here if `userId === session.user.id` and if the user is authorized to access the data of this teamId? Or am I missing something and this is done somewhere else?

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
Tested and works, but please address my comments first 🙏

File: apps/web/pages/api/insights/download.ts (hunk 10 above)
Comment (alannnc) [reply apps/web/pages/api/insights/download.ts:56]: You are right, I forgot about these validations since I cannot use trpc procedures here :/

**Comment by alannnc** [comment]
I'm gonna try the trpc approach while also adding public url link for this PR.

**Review by CarinaWolli — APPROVED** [review]
Looks good, my requested changes are fixed 👍

**Pull request merged**